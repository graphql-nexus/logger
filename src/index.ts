export { demo } from './demo'
export { Logger } from './logger'
export { create, RootLogger } from './root-logger'
export { Data as SettingsData, Input as SettingsInput } from './settings'
