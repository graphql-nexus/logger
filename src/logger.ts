import * as Lo from 'lodash'
import * as OS from 'os'
import { Level, LevelNum, LEVELS } from './level'
import * as Prettifier from './prettifier'
import * as RootLogger from './root-logger'

// TODO JSON instead of unknown type
type Context = Record<string, unknown>

export type LogRecord = {
  level: LevelNum
  path: string[]
  event: string
  context: Context
  time: number
  pid: number
  hostname: string
}

type Log = (event: string, context?: Context) => void

export type Logger = {
  fatal: Log
  error: Log
  warn: Log
  info: Log
  debug: Log
  trace: Log
  addToContext: (context: Context) => Logger // fluent
  child: (name: string) => Logger // fluent
}

/**
 * Create a logger.
 */
export function create(
  rootState: RootLogger.State,
  path: string[],
  parentContext: Context
): { logger: Logger; link: Link } {
  const state: State = {
    // Copy as addToContext will mutate it
    pinnedAndParentContext: Lo.cloneDeep(parentContext),
    children: [],
  }

  function updateContextAndPropagate(newContext: Context) {
    state.pinnedAndParentContext = newContext
    state.children.forEach((child) => {
      child.onNewParentContext(state.pinnedAndParentContext)
    })
  }

  function send(level: Level, event: string, localContext: undefined | Context) {
    const thisLevel = LEVELS[level].number
    const levelSetting = LEVELS[rootState.settings.level].number
    if (thisLevel >= levelSetting) {
      // Avoid mutating the passed local context
      const context = localContext
        ? Lo.merge({}, state.pinnedAndParentContext, localContext)
        : state.pinnedAndParentContext
      const logRec: LogRecord = {
        path,
        context,
        event,
        level: thisLevel,
        time: Date.now(),
        hostname: OS.hostname(),
        pid: process.pid,
      }
      const logMsg = rootState.settings.pretty.enabled
        ? Prettifier.render(rootState.settings.pretty, logRec)
        : JSON.stringify(logRec)
      rootState.settings.output.write(logMsg + OS.EOL)
    }
  }

  const link: Link = {
    onNewParentContext(newParentContext: Context) {
      updateContextAndPropagate(
        Lo.merge(
          // Copy so that we don't mutate parent while maintaining local overrides...
          {},
          newParentContext,
          // ...this
          state.pinnedAndParentContext
        )
      )
    },
  }

  const logger: Logger = {
    fatal(event, context) {
      send('fatal', event, context)
    },
    error(event, context) {
      send('error', event, context)
    },
    warn(event, context) {
      send('warn', event, context)
    },
    info(event, context) {
      send('info', event, context)
    },
    debug(event, context) {
      send('debug', event, context)
    },
    trace(event, context) {
      send('trace', event, context)
    },
    addToContext(context: Context) {
      // Can safely mutate here, save some electricity...
      updateContextAndPropagate(Lo.merge(state.pinnedAndParentContext, context))
      return logger
    },
    child: (name: string): Logger => {
      const { logger: child, link } = create(rootState, path.concat([name]), state.pinnedAndParentContext)
      state.children.push(link)
      return child
    },
  }

  return {
    logger,
    link,
  }
}

type Link = {
  onNewParentContext: (newContext: Context) => void
}

type State = {
  pinnedAndParentContext: Context
  children: Link[]
}
