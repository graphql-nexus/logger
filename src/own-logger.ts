import * as Logger from './'

export const log = Logger.create({ name: 'nexusLogger' })
