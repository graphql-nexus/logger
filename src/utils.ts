import { format } from 'util'
/**
 * Guarantee the length of a given string, padding before or after with the
 * given character. If the given string is longer than  the span target, then it
 * will be cropped.
 */
export function span(
  padSide: 'padBefore' | 'padAfter',
  padChar: string,
  target: number,
  content: string
): string {
  if (content.length > target) {
    return content.slice(0, target)
  }
  let toPadSize = target - content.length
  while (toPadSize > 0) {
    if (padSide === 'padAfter') {
      content = content + padChar
    } else if (padSide === 'padBefore') {
      content = padChar + content
    }
    toPadSize--
  }
  return content
}

/**
 * Guarantee the length of a given string, padding with space as needed. Content
 * is aligned left and if exceeds span target length to begin with gets cropped.
 */
export const clampSpace = span.bind(null, 'padAfter', ' ')

/**
 * Create a string of space of the given length.
 */
export function spanSpace(num: number): string {
  return spanChar(num, ' ')
}

/**
 * Create a string of the given length and character
 */
export function spanChar(num: number, char: string): string {
  return range(num).map(constant(char)).join('')
}

/**
 * Guarantee the length of a given string, padding with space as needed. Content
 * is aligned right and if exceeds span target length to begin with gets cropped.
 */
export const spanSpaceRight = span.bind(null, 'padBefore', ' ')

/**
 * Use this to make assertion at end of if-else chain that all members of a
 * union have been accounted for.
 */
export function casesHandled(x: never): never {
  throw new Error(`A case of value was not handled: ${x}`)
}

/**
 * Create a function that will only ever return the given value when called.
 */
export function constant<T>(x: T): () => T {
  return function () {
    return x
  }
}

/**
 * Create a range of integers.
 */
export function range(times: number): number[] {
  const list: number[] = []
  while (list.length < times) {
    list.push(list.length + 1)
  }
  return list
}

/**
 * Strip keys from object whose value is undefined.
 */
export function omitUndefinedKeys<T extends Record<string, unknown>>(data: T): T {
  return Object.entries(data ?? {})
    .filter(([k, v]) => v !== undefined)
    .reduce((acc, [k, v]) => Object.assign(acc, { [k]: v }), {} as T)
}

/**
 * Get the last item of an array.
 */
export function last<T>(xs: T[]): T {
  return xs[xs.length - 1]
}

export function isEmpty(x?: object): boolean {
  if (x === undefined) return true
  return Object.values(x).filter((val) => val !== undefined).length === 0
}

/**
 * Run a given parser over an environment variable. If parsing fails, throw a
 * contextual error message.
 */
export function parseFromEnvironment<T>(
  key: string,
  parser: {
    info: { valid: string; typeName: string }
    run: (raw: string) => null | T
  }
): T {
  const envVarValue = process.env[key]! // assumes env presence handled before
  const result = parser.run(envVarValue)

  if (result === null) {
    throw new Error(
      `Could not parse environment variable ${key} into ${
        parser.info.typeName
      }. The environment variable was: ${format(envVarValue)}. A valid environment variable must be like: ${
        parser.info.valid
      }`
    )
  }

  return result
}
