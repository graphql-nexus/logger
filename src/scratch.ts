import { chalk } from './chalk'
import { validPathSegmentNameRegex } from './logger'

console.log('')
console.log('')
console.log(renderSyntaxManual())


