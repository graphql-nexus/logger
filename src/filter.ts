import { Either, isLeft, isRight, left, right } from 'fp-ts/lib/Either'
import { chalk } from './chalk'
import { validPathSegmentNameRegex } from './data'
import * as Level from './level'
import type { LogRecord } from './logger'
import { casesHandled, ContextualError, createContextualError, getLeft, last, rightOrThrow } from './utils'

// https://regex101.com/r/6g6BHc/1
// If you change the regex please update the link and vice-versa!
const pathTerminalRegex = /^([A-z_]*)(\*)?(?:(?:@(1|2|3|4|5|6|trace|debug|info|warn|error|fatal)([-+]?))|(@\*))?$/

const symbols = {
  negate: '!',
  pathDelim: ':',
  patternDelim: ',',
  descendents: '*',
  levelDelim: '@',
  levelGte: '+',
  levelLte: '-',
}

export type Parsed = {
  originalInput: string
  level: {
    value: Level.Name | '*'
    comp: 'lte' | 'gte' | 'eq'
  }
  negate: boolean
  path: {
    value: string
    descendents:
      | false
      | {
          includeParent: boolean
        }
  }
}

/**
 * Some of the criteria a pattern can specify are optional. When such criteria
 * are not specified, then these defaults are used.
 */
export type Defaults = {
  level: {
    value: Level.Name
    comp: Parsed['level']['comp']
  }
}

/**
 * Parse a full pattern. This accounts for lists of patterns. This is the parsing entrypoint.
 */
export function parse(defaults: Defaults, pattern: string): Either<ParseError, Parsed>[] {
  // Allow sloppy lists so long as there is at least one pattern given
  const patterns = pattern
    .split(symbols.patternDelim)
    .map((p) => p.trim())
    .filter((p) => p !== '')

  // if (!patterns.length) {
  // createInvalidPattern(pattern, 'There must be at least one pattern present.')
  // }

  return patterns.map((p) => parseOne(defaults, p))
}

/**
 * Parse a single pattern. This assumes parsing of "," has already been handled
 * including whitespace trimming around the pattern.
 */
export function parseOne(criteriaDefaults: Defaults, pattern: string): Either<ParseError, Parsed> {
  let pattern_ = pattern
  // todo maybe level default should be wildcard instead...
  let level = { ...criteriaDefaults.level } as Parsed['level']
  let negate = false
  let path: Parsed['path'] = { value: '', descendents: false }

  // process negate

  if (pattern_[0] === '!') {
    negate = true
    pattern_ = pattern_.slice(1)
  }

  // process everything else

  let pathParts = pattern_.split(symbols.pathDelim)

  if (last(pathParts).match(validPathSegmentNameRegex)) {
    path.value = pathParts.join(symbols.pathDelim)
  } else {
    const match = last(pathParts).match(pathTerminalRegex)
    if (match) {
      type LevelNumString = '1' | '2' | '3' | '4' | '5' | '6'
      pathParts.pop()
      const pathPart = match[1] as undefined | string
      const descendents = match[2] as undefined | string
      const levelValue = match[3] as undefined | Level.Name | LevelNumString
      const levelDir = match[4] as undefined | '-' | '+'
      const levelWildCard = match[5] as undefined | string

      if (pathPart) {
        path.value = pathParts.concat([pathPart]).join(symbols.pathDelim)
      } else {
        path.value = pathParts.join(symbols.pathDelim)
      }

      if (descendents) {
        path.descendents = {
          includeParent: Boolean(pathPart),
        }
      }

      if (levelValue) {
        // the original regex guarantees 1-5 so we don't have to validate that now
        if (levelValue.match(/\d/)) {
          level.value = Level.LEVELS_BY_NUM[levelValue as LevelNumString].label
        } else {
          level.value = levelValue as Level.Name
        }
        if (levelDir === '+') level.comp = 'gte'
        else if (levelDir === '-') level.comp = 'lte'
        else if (levelDir === '') level.comp = 'eq'
        else if (levelDir === undefined) level.comp = 'eq'
        else casesHandled(levelDir)
      } else if (levelWildCard) {
        level.value = '*'
        level.comp = 'eq'
      }
    } else {
      return left(createInvalidPattern(pattern))
    }
  }

  // check for errors

  if (path.value !== '') {
    const invalidPathPartNames = path.value
      .split(symbols.pathDelim)
      .filter((pathPart) => !pathPart.match(validPathSegmentNameRegex))

    if (invalidPathPartNames.length) {
      return left(
        createInvalidPattern(
          pattern,
          `Path segment names must only contain ${String(validPathSegmentNameRegex)}.`
        )
      )
    }
  } else if (path.value === '' && !path.descendents) {
    return left(createInvalidPattern(pattern))
  }

  return right({
    negate,
    path,
    originalInput: pattern,
    level,
  })
}

/**
 * Test if a log matches the pattern.
 */
export function test(patterns: Readonly<Parsed[]>, log: LogRecord): boolean {
  let yaynay = false
  for (const pattern of patterns) {
    // if log already passed then we can skip rest except negations
    if (yaynay && !pattern.negate) continue

    const logPath = log.path?.join(symbols.pathDelim) ?? null
    let isPass = false

    // test in order of computational cost, short-citcuiting ASAP

    // level

    if (pattern.level.value === '*') {
      isPass = true
    } else {
      isPass = comp(pattern.level.comp, log.level, Level.LEVELS[pattern.level.value].number)
    }

    // path

    if (isPass) {
      if (pattern.path.descendents) {
        if (logPath === pattern.path.value) {
          isPass = pattern.path.descendents.includeParent
        } else if (logPath === null) {
          isPass = false
        } else {
          isPass = logPath.startsWith(pattern.path.value)
        }
      } else {
        isPass = pattern.path.value === logPath
      }
    }

    // Allow negates to undo previous passes while non-negates can only pass,
    // not unpass.
    if (pattern.negate) {
      yaynay = !isPass
    } else if (isPass) {
      yaynay = true
    }
  }

  return yaynay
}

function comp(kind: Parsed['level']['comp'], a: number, b: number): boolean {
  if (kind === 'eq') return a === b
  if (kind === 'gte') return a >= b
  if (kind === 'lte') return a <= b
  casesHandled(kind)
}

/**
 * Like `parse` but throws upon any failure.
 *
 * @remarks
 *
 * Only use this if you know what you're doing.
 */
export function parseUnsafe(defaults: Defaults, pattern: string): Parsed[] {
  return parse(defaults, pattern).map(rightOrThrow)
}

type ParseError = ContextualError<{ pattern: string; hint?: string }>

function createInvalidPattern(pattern: string, hint?: string): ParseError {
  return createContextualError(`Invalid filter pattern: "${pattern}"`, { pattern, hint })
}

/**
 * Get the string contents of a manual showing how to write filters.
 */
export function renderSyntaxManual() {
  const m = chalk.magenta
  const b = chalk.blue
  const gray = chalk.gray
  const c = chalk.cyan
  const bold = chalk.bold
  const subtle = (x: string) => gray(x)
  const subtitle = (x: string) => bold(m(x))
  const pipe = gray('|')
  return `${bold(b('LOG FILTERING SYNTAX MANUAL  ⟁'))}
${bold(b(`▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬`))}

${bold(b(`Grammar`))}

    [!](${c(`<path>`)}|*)[@((${c(`<levelNum>`)}|${c(`<levelLabel>`)})[+-]|*)][,<...>]

    ${c(`<path>`)}       = ${validPathSegmentNameRegex.toString().replace(/(^\/|\/$)/g, gray(`$1`))}
    ${c(`<levelNum>`)}   = 1     ${pipe} 2     ${pipe} 3    ${pipe} 4    ${pipe} 5     ${pipe} 6
    ${c(`<levelLabel>`)} = trace ${pipe} debug ${pipe} info ${pipe} warn ${pipe} error ${pipe} fatal

${bold(b(`Examples`))}

    ${subtitle(`Paths`)}
    app         ${subtle(`app path at default level`)}
    app:router  ${subtle(`app router path at default level`)} 

    ${subtitle(`Wildcards Paths`)}
    *           ${subtle(`all paths at default level`)} 
    app:*       ${subtle(`descendent paths of app at defualt level`)}
    app*        ${subtle(`app path and its descendent paths at defualt level`)}

    ${subtitle(`Negation`)}
    !app        ${subtle(`all paths expect app at default level`)} 

    ${subtitle(`Lists`)}
    app,nexus   ${subtle(`app and nexus paths at default level`)}

    ${subtitle(`Levels`)}
    *@info      ${subtle(`all paths at info level`)}
    *@error-    ${subtle(`all paths at error level or lower`)}
    *@debug+    ${subtle(`all paths at debug level or higher`)}
    *@3         ${subtle(`all paths at info level`)}
    *@4-        ${subtle(`all paths at error level or lower`)}
    *@2+        ${subtle(`all paths at debug level or higher`)}

    ${subtitle(`Wildcard Paths`)}
    *@*         ${subtle(`all paths at all levels`)}

    ${subtitle(`Mixed`)}
    app@*       ${subtle(`app path at all levels`)}
    app:*@2+    ${subtle(`descendent paths of app at debug level or higher`)}
    app*@2-     ${subtle(`app & descendent paths of app at debug level or lower`)} 
  `
}

export function renderSyntaxError(input: {
  errPatterns: Either<ParseError, Parsed>[]
  foundIn?: string
  some?: boolean
}): string {
  const badOnes = input.errPatterns.filter(isLeft)
  const multipleInputs = input.errPatterns.length > 1
  const multipleErrors = badOnes.length > 1
  const allBad = badOnes.length === input.errPatterns.length
  const foundIn = `${input.foundIn ? ` found in ${input.foundIn}` : ''}`
  let message

  if (!multipleInputs) {
    const e = getLeft(badOnes[0])
    const pattern = e?.context.pattern
    const hint = e?.context.hint ? `. ${e.context.hint}` : ''
    message = `Your log filter's pattern${foundIn} was invalid: "${chalk.red(
      pattern
    )}${hint}"\n\n${renderSyntaxManual()}`
  } else if (!multipleErrors) {
    const e = getLeft(badOnes[0])
    const pattern = e?.context.pattern
    const hint = e?.context.hint ? `. ${e.context.hint}` : ''
    message = `One of the patterns in your log filter${foundIn} was invalid: "${chalk.red(
      pattern
    )}"${hint}\n\n${renderSyntaxManual()}`
  } else {
    const patterns = badOnes
      .map((e) => {
        const hint = e.left.context.hint ? chalk.gray(`  ${e.left.context.hint}`) : ''
        return `    ${chalk.red(e.left.context.pattern)}${hint}`
      })
      .join('\n')
    const intro = allBad
      ? `All of the patterns in your log filter`
      : `Some (${badOnes.length}) of the patterns in your log filter`
    message = `${intro}${foundIn} were invalid:\n\n${patterns}\n\n${renderSyntaxManual()}`
  }

  return message
}

export function processLogFilterInput(
  defaults: Defaults,
  pattern: string,
  foundIn?: string
): null | Parsed[] {
  const errPatterns = parse(defaults, pattern)
  const goodOnes = errPatterns.filter(isRight)
  const badOnes = errPatterns.filter(isLeft)
  let patterns = null

  if (badOnes.length) {
    if (goodOnes.length) {
      patterns = goodOnes.map(rightOrThrow)
    }
    const message = renderSyntaxError({ errPatterns, foundIn })
    // todo use logger
    // but doing so introduces circ dependency breaking cjs output
    // try ts 3.9 live bindings output?
    // no problem in esm-world
    // log.warn(message)
    console.log(message)
  } else {
    patterns = errPatterns.map(rightOrThrow)
  }

  return patterns
}
