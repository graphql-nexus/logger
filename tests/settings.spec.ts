import * as Logger from '../src'
import { createMockOutput, MockOutput, resetBeforeEachTest } from './__helpers'

let log: Logger.RootLogger
let output: MockOutput

resetBeforeEachTest(process, 'env')

beforeEach(() => {
  process.env.LOG_PRETTY = 'false'
  output = createMockOutput()
  log = Logger.create({ output, pretty: { timeDiff: false } })
  process.stdout.columns = 200
})

describe('pretty', () => {
  describe('.enabled', () => {
    it('can be disabled', () => {
      expect(Logger.create({ pretty: { enabled: false } }).settings.pretty.enabled).toEqual(false)
    })
    it('persists across peer field changes', () => {
      const l = Logger.create({ pretty: { enabled: false } })
      l.settings({ pretty: { color: false } })
      expect(l.settings.pretty).toEqual({
        enabled: false,
        color: false,
        levelLabel: false,
        timeDiff: true,
      })
    })
    // it('controls if logs are rendered pretty or as JSON', () => {
    //   if (!process.version.match(/^v12/)) return
    //   output.captureConsoleLog()
    //   log.info('foo')
    //   Logger.demo(log)
    //   expect(output.memory.jsonOrRaw).toMatchSnapshot()
    // })
    describe('precedence', () => {
      it('considers instnace time config first', () => {
        process.stdout.isTTY = false
        process.env.LOG_PRETTY = 'false'
        const l = Logger.create({ pretty: false })
        l.settings({ pretty: true })
        expect(l.settings.pretty.enabled).toEqual(true)
      })
      it('then considers construction time config', () => {
        process.stdout.isTTY = false
        process.env.LOG_PRETTY = 'false'
        const l = Logger.create({ pretty: true })
        expect(l.settings.pretty.enabled).toEqual(true)
      })
      it('then considers LOG_PRETTY env var true|false (case insensitive)', () => {
        process.stdout.isTTY = false
        process.env.LOG_PRETTY = 'tRuE'
        const l = Logger.create()
        expect(l.settings.pretty.enabled).toEqual(true)
      })
      it('then defaults to process.stdout.isTTY', () => {
        delete process.env.LOG_PRETTY // pre-test logic forces false otherwise
        process.stdout.isTTY = true
        const l = Logger.create()
        expect(l.settings.pretty.enabled).toEqual(true)
      })
    })
  })

  describe('.color', () => {
    it('controls if pretty logs have color or not', () => {
      log.settings({ pretty: { enabled: true, color: false } })
      log.info('foo', { qux: true })
      expect(output.memory.jsonOrRaw).toMatchSnapshot()
    })
    it('can be disabled', () => {
      expect(Logger.create({ pretty: { enabled: false, color: false } }).settings.pretty.color).toEqual(false)
    })
    it('is true by default', () => {
      expect(Logger.create({ pretty: { enabled: true } }).settings.pretty).toEqual({
        enabled: true,
        color: true,
        levelLabel: false,
        timeDiff: true,
      })
      expect(Logger.create({ pretty: { enabled: false } }).settings.pretty).toEqual({
        enabled: false,
        color: true,
        levelLabel: false,
        timeDiff: true,
      })
    })
    it('persists across peer field changes', () => {
      const l = Logger.create({
        pretty: { enabled: false, color: false },
      })
      l.settings({ pretty: { enabled: true } })
      expect(l.settings.pretty).toEqual({
        enabled: true,
        color: false,
        levelLabel: false,
        timeDiff: true,
      })
    })
  })
  describe('.levelLabel', () => {
    it('is false by default', () => {
      expect(Logger.create().settings.pretty.levelLabel).toEqual(false)
    })
    it('can be enabled', () => {
      expect(Logger.create({ pretty: { levelLabel: true } }).settings.pretty.levelLabel).toEqual(true)
      expect(Logger.create().settings({ pretty: { levelLabel: true } }).settings.pretty.levelLabel).toEqual(
        true
      )
    })
    it('persists across peer field changes', () => {
      const l = Logger.create({ pretty: { levelLabel: true } })
      l.settings({ pretty: false })
      expect(l.settings.pretty.levelLabel).toBe(true)
    })
    it('controls if label is spelt out or not', () => {
      log.settings({
        pretty: { enabled: true, levelLabel: true, color: false },
      })
      log.fatal('foo')
      log.error('foo')
      log.warn('foo')
      log.info('foo')
      log.debug('foo')
      log.trace('foo')
      expect(output.memory.jsonOrRaw).toMatchSnapshot()
    })
  })
  describe('.timeDiff', () => {
    it('is true by default', () => {
      expect(Logger.create().settings.pretty.timeDiff).toEqual(true)
    })
    it('can be disabled', () => {
      const l1 = Logger.create({ pretty: { timeDiff: false } })
      expect(l1.settings.pretty.timeDiff).toEqual(false)
      const l2 = Logger.create()
      l2.settings({ pretty: { timeDiff: false } })
      expect(l2.settings.pretty.levelLabel).toEqual(false)
    })
    it('persists across peer field changes', () => {
      const l = Logger.create({ pretty: { timeDiff: false } })
      l.settings({ pretty: false })
      expect(l.settings.pretty.timeDiff).toBe(false)
    })
    it('controls presence of time deltas in gutter', () => {
      log.settings({
        pretty: { enabled: true, color: false, timeDiff: true },
      })
      log.info('a') // prep the next delta, this too unreliable to test
      log.info('b')
      log.info('c')
      const pattern = /^  (?:\d| )\d.*/ // non-deterministic timing here, allow for 1-2 digit ms timings
      expect(output.memory.jsonOrRaw[1]).toMatch(pattern)
      expect(output.memory.jsonOrRaw[2]).toMatch(pattern)
    })
    // todo these tests as unit level to some pure logic functions would be
    // easy... e.g. prettifier.spec.ts ... But then we run the risk of sliding
    // toward testing internals too much :\
    it.todo('renders as secodns if >= 10s')
    it.todo('renders as minutes if >= 100s')
    it.todo('renders as hours if >= 60m')
    it.todo('renders as days if >= 24h')
  })

  describe('shorthands', () => {
    it('true means enabled true', () => {
      expect(Logger.create({ pretty: true }).settings.pretty).toEqual({
        enabled: true,
        color: true,
        levelLabel: false,
        timeDiff: true,
      })
    })

    it('false means enabled false', () => {
      expect(Logger.create({ pretty: false }).settings.pretty).toEqual({
        enabled: false,
        color: true,
        levelLabel: false,
        timeDiff: true,
      })
    })
  })
})

describe('level', () => {
  describe('precedence', () => {
    it('considers instance time config first', () => {
      process.env.NODE_ENV = 'production'
      process.env.LOG_LEVEL = 'fatal'
      const l = Logger.create({ level: 'fatal' })
      l.settings({ level: 'trace' })
      expect(l.settings.level).toEqual('trace')
    })

    it('then considers construction time config', () => {
      process.env.NODE_ENV = 'production'
      process.env.LOG_LEVEL = 'fatal'
      const l = Logger.create({ level: 'trace' })
      expect(l.settings.level).toEqual('trace')
    })

    it('then considers LOG_LEVEL env var', () => {
      process.env.NODE_ENV = 'production'
      process.env.LOG_LEVEL = 'trace'
      const l = Logger.create()
      expect(l.settings.level).toEqual('trace')
    })

    it('then considers NODE_ENV=production', () => {
      process.env.NODE_ENV = 'production'
      const l = Logger.create()
      expect(l.settings.level).toEqual('info')
    })

    it('then defaults to debug', () => {
      const l = Logger.create()
      expect(l.settings.level).toEqual('debug')
    })
  })

  it('logs below set level are not output', () => {
    log.settings({ level: 'warn' }).info('foo')
    expect(output.memory.jsonOrRaw).toEqual([])
  })

  it('LOG_LEVEL env var config is treated case insensitive', () => {
    process.env.NODE_ENV = 'production'
    process.env.LOG_LEVEL = 'TRACE'
    const l = Logger.create()
    expect(l.settings.level).toEqual('trace')
  })

  it('LOG_LEVEL env var config when invalid triggers thrown readable error', () => {
    process.env.LOG_LEVEL = 'ttrace'
    expect(() => Logger.create()).toThrowErrorMatchingInlineSnapshot(
      `"Could not parse environment variable LOG_LEVEL into LogLevel. The environment variable was: ttrace. A valid environment variable must be like: fatal, error, warn, info, debug, trace"`
    )
  })
})

describe('data', () => {
  it('all defaults to false if process.NODE_ENV is not production', () => {
    expect(Logger.create().settings.data).toMatchInlineSnapshot(`
      Object {
        "hostname": false,
        "pid": false,
        "time": false,
      }
    `)
    process.env.NODE_ENV = 'not production'
    expect(Logger.create().settings.data).toMatchInlineSnapshot(`
      Object {
        "hostname": false,
        "pid": false,
        "time": false,
      }
    `)

    process.env.NODE_ENV = 'production'
    expect(Logger.create().settings.data).toMatchInlineSnapshot(`
      Object {
        "hostname": true,
        "pid": true,
        "time": true,
      }
    `)
  })

  it('merges initial setting with defaults', () => {
    expect(Logger.create({ data: { time: true } }).settings.data).toMatchInlineSnapshot(`
      Object {
        "hostname": false,
        "pid": false,
        "time": true,
      }
    `)
    process.env.NODE_ENV = 'not production'
    expect(Logger.create({ data: { time: true } }).settings.data).toMatchInlineSnapshot(`
      Object {
        "hostname": false,
        "pid": false,
        "time": true,
      }
    `)
    process.env.NODE_ENV = 'production'
    expect(Logger.create({ data: { time: false } }).settings.data).toMatchInlineSnapshot(`
      Object {
        "hostname": true,
        "pid": true,
        "time": false,
      }
    `)
  })

  it('merges incremental changes with previous state', () => {
    expect(Logger.create().settings({ data: { time: true } }).settings.data).toMatchInlineSnapshot(`
      Object {
        "hostname": false,
        "pid": false,
        "time": true,
      }
    `)
    process.env.NODE_ENV = 'not production'
    expect(Logger.create().settings({ data: { time: true } }).settings.data).toMatchInlineSnapshot(`
      Object {
        "hostname": false,
        "pid": false,
        "time": true,
      }
    `)
    process.env.NODE_ENV = 'production'
    expect(Logger.create().settings({ data: { time: false } }).settings.data).toMatchInlineSnapshot(`
      Object {
        "hostname": true,
        "pid": true,
        "time": false,
      }
    `)
  })
})
