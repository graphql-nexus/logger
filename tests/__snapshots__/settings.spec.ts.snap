// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter LOG_FILTER envar config when invalid triggers readable log warning 1`] = `
Array [
  Array [
    "Your log filter's pattern found in environment variable LOG_FILTER. was invalid: \\"**\\"

LOG FILTERING SYNTAX MANUAL  ⟁
▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬

Grammar

    Precise: 

    /^(!)?((?:[A-z_]+:)+)?(?:((:)?\\\\*)?|([A-z_]+|))?(?<=\\\\*|[A-z_]+)(?:(?:@(1|2|3|4|5|6|trace|debug|info|warn|error|fatal)([-+]?))|(@\\\\*))?$/

    Generally: 

    [!](*|<path>[:*])[@(*|(<levelNum>|<levelLabel>)[+-])] [,<pattern>]

    <path>       = /^[A-z_]+[A-z_0-9]*$/ [:<path>]
    <levelNum>   = 1     | 2     | 3    | 4    | 5     | 6
    <levelLabel> = trace | debug | info | warn | error | fatal

Examples

    Paths
    app         app path at default level
    app:router  app router path at default level 

    Wildcards Paths
    *           all paths at default level 
    app:*       app path and its descendant paths at defualt level
    app::*      descendant paths of app at defualt level

    Negation
    !app        all paths expect app at default level 

    Lists
    app,nexus   app and nexus paths at default level

    Levels
    *@info      all paths at info level
    *@error-    all paths at error level or lower
    *@debug+    all paths at debug level or higher
    *@3         all paths at info level
    *@4-        all paths at error level or lower
    *@2+        all paths at debug level or higher

    Wildcard Paths
    *@*         all paths at all levels

    Mixed
    app@*       app path at all levels
    app:*@2+    descendant paths of app at debug level or higher
    app*@2-     app & descendant paths of app at debug level or lower 
  ",
  ],
]
`;

exports[`filter precedence considers instance time config first 1`] = `
Array [
  Object {
    "level": Object {
      "comp": "gte",
      "value": "debug",
    },
    "negate": false,
    "originalInput": "foo",
    "path": Object {
      "descendants": false,
      "value": ".:foo",
    },
  },
]
`;

exports[`filter precedence then considers LOG_FILTER env var 1`] = `
Object {
  "defaults": Object {
    "level": Object {
      "comp": "gte",
      "value": "debug",
    },
  },
  "originalInput": undefined,
  "patterns": Array [
    Object {
      "level": Object {
        "comp": "gte",
        "value": "debug",
      },
      "negate": false,
      "originalInput": "from_env_var",
      "path": Object {
        "descendants": false,
        "value": ".:from_env_var",
      },
    },
  ],
}
`;

exports[`filter precedence then considers construction time config 1`] = `
Object {
  "defaults": Object {
    "level": Object {
      "comp": "gte",
      "value": "debug",
    },
  },
  "originalInput": "*@fatal",
  "patterns": Array [
    Object {
      "level": Object {
        "comp": "eq",
        "value": "fatal",
      },
      "negate": false,
      "originalInput": "*@fatal",
      "path": Object {
        "descendants": Object {
          "includeParent": true,
        },
        "value": ".",
      },
    },
  ],
}
`;

exports[`filter precedence then defaults to "* 1`] = `
Object {
  "defaults": Object {
    "level": Object {
      "comp": "gte",
      "value": "debug",
    },
  },
  "originalInput": "*",
  "patterns": Array [
    Object {
      "level": Object {
        "comp": "gte",
        "value": "debug",
      },
      "negate": false,
      "originalInput": "*",
      "path": Object {
        "descendants": Object {
          "includeParent": true,
        },
        "value": ".",
      },
    },
  ],
}
`;

exports[`pretty .color controls if pretty logs have color or not 1`] = `
Array [
  "● foo  --  qux: true
",
]
`;

exports[`pretty .levelLabel controls if label is spelt out or not 1`] = `
Array [
  "✕ fatal foo
",
  "■ error foo
",
  "▲ warn  foo
",
  "● info  foo
",
  "○ debug foo
",
]
`;
